internal/engine/docker_compose_build_and_deployer.go

```diff
@@ -12,58 +14,92 @@ import (
-func (bd *DockerComposeBuildAndDeployer) extract(specs []model.TargetSpec) []model.DockerComposeTarget {
-	result := []model.DockerComposeTarget{}
+// Extract the targets we can apply -- DCBaD supports ImageTargets and DockerComposeTargets.
+func (bd *DockerComposeBuildAndDeployer) extract(specs []model.TargetSpec) ([]model.ImageTarget, []model.DockerComposeTarget) {
+	var iTargets []model.ImageTarget
+	var dcTargets []model.DockerComposeTarget
+
 	for _, s := range specs {
-		dc, isDC := s.(model.DockerComposeTarget)
-		if !isDC {
-			return nil
+		switch s := s.(type) {
+		case model.ImageTarget:
+			iTargets = append(iTargets, s)
+		case model.DockerComposeTarget:
+			dcTargets = append(dcTargets, s)
 		}
-		result = append(result, dc)
 	}
-	return result
+	return iTargets, dcTargets
 }
 
 func (bd *DockerComposeBuildAndDeployer) BuildAndDeploy(ctx context.Context, specs []model.TargetSpec, currentState store.BuildStateSet) (store.BuildResultSet, error) {
-	dcs := bd.extract(specs)
-	if len(dcs) == 0 {
-		return store.BuildResultSet{}, RedirectToNextBuilderf("Specs not supported by DockerComposeBuildAndDeployer")
+	iTargets, dcTargets := bd.extract(specs)
+	if len(dcTargets) != 1 {
+		return store.BuildResultSet{}, RedirectToNextBuilderf(
+			"DockerComposeBuildAndDeployer requires exactly one dcTarget (got %d)", len(dcTargets))
+	}
+	if len(iTargets) > 1 {
+		return store.BuildResultSet{}, RedirectToNextBuilderf(
+			"DockerComposeBuildAndDeployer supports at most one ImageTarget (got %d)", len(iTargets))
 	}
+	dcTarget := dcTargets[0]
+	haveImage := len(iTargets) == 1
+
 	span, ctx := opentracing.StartSpanFromContext(ctx, "DockerComposeBuildAndDeployer-BuildAndDeploy")
-	span.SetTag("target", dcs[0].Name)
+	span.SetTag("target", dcTargets[0].Name)
 	defer span.Finish()
 
-	brs := store.BuildResultSet{}
-	stdout := logger.Get(ctx).Writer(logger.InfoLvl)
-	stderr := logger.Get(ctx).Writer(logger.InfoLvl)
-	for _, dc := range dcs {
-		err := bd.dcc.Up(ctx, dc.ConfigPath, dc.Name, stdout, stderr)
-		if err != nil {
-			return store.BuildResultSet{}, err
-		}
+	if haveImage {
+		logger.Get(ctx).Infof("~~ i'm building ur image for: %s", dcTarget.Name)
+		var err error
+		var ref reference.NamedTagged
+		results := store.BuildResultSet{}
+		iTarget := iTargets[0]
+
+		ps := build.NewPipelineState(ctx, 1, bd.ibd.clock)
+		defer func() { ps.End(ctx, err) }()
 
-		// NOTE(dmiller): right now we only need this the first time. In the future
-		// it might be worth it to move this somewhere else
-		cid, err := bd.dcc.ContainerID(ctx, dc.ConfigPath, dc.Name)
+		// NOTE(maia): we assume that this func takes one DC target and up to one image target
+		// corresponding to that service. If this func ever supports specs for more than one
+		// service at once, we'll have to match up image build results to DC target by ref.
+		ref, err = bd.ibd.build(ctx, iTarget, currentState[iTarget.ID()], ps, true)
```

Comment:
maybe this is a good sign that this needs to be broken out into a common utility class? it seems very weird, architecturally, for one ImageBuildAndDeployer to container another.

Reply:
will do